src/routing/error_cause.rs

```diff
@@ -234,9 +217,101 @@ impl From<anyhow::Error> for ErrorCause {
+            Self::UnknownDomain => StatusCode::BAD_REQUEST,
+            Self::UpstreamError => StatusCode::SERVICE_UNAVAILABLE,
+        }
+    }
+
+    pub fn details(&self) -> String {
+        match self {
+            Self::BodyTimedOut => "Reading the request body timed out due to data arriving too slowly.".to_string(),
+            Self::CanisterIdNotFound => "The canister ID could not be resolved from the provided authority.".to_string(),
+            Self::Denylisted => "Access to this resource is denied due to a violation of the code of conduct.".to_string(),
+            Self::DomainCanisterMismatch => "Access to the canister is forbidden through the current gateway domain. Try accessing it through an allowed gateway domain.".to_string(),
+            Self::IncorrectPrincipal => "The principal in the request is incorrectly formatted.".to_string(),
+            Self::LoadShed => "The HTTP gateway is temporarily unable to handle the request due to high load. Please try again later.".to_string(),
+            Self::MalformedRequest(x) => x.to_string(),
+            Self::NoAuthority => "The request is missing the required authority information (e.g., Host header).".to_string(),
+            Self::Other => "Internal Server Error".to_string(),
+            Self::PayloadTooLarge => "The payload is too large.".to_string(),
+            Self::RateLimited => "Rate limit exceeded. Please slow down requests and try again later.".to_string(),
+            Self::SubnetUnavailable => "The subnet is temporarily unavailable. This may be due to an ongoing upgrade of the replica software. Please try again later.".to_string(),
+            Self::UnknownDomain => "The requested domain is not served by this HTTP gateway.".to_string(),
+            Self::UpstreamError => "The HTTP gateway is temporarily unable to process the request. Please try again later.".to_string(),
+        }
+    }
+
+    pub fn html(&self) -> String {
+        match self {
+            Self::Denylisted => include_str!("error_pages/451.html").to_string(),
+            _ => {
+                let template = ERROR_PAGE_TEMPLATE;
+                let template = template.replace("{status_code}", self.status_code().as_str());
+                let template = template.replace("{reason}", self.to_string().as_str());
+                let template = template.replace("{details}", self.details().as_str());
+                template
+            }
+        }
+    }
+}
+
+// Creates the response from ErrorClientFacing
+impl IntoResponse for ErrorClientFacing {
+    fn into_response(self) -> Response {
+        // Return an HTML error page if it was an HTTP request
+        let body = match ERROR_CONTEXT.get() {
+            RequestType::Http => format!("{}\n", self.html()),
+            _ => format!("error: {}\ndetails: {}", self.to_string(), self.details()),
+        };
+
+        let mut resp = (self.status_code(), body).into_response();
+        if ERROR_CONTEXT.get() == RequestType::Http {
```

Comment:
Maybe avoid doing .get() multiple times here? just assign it to some var once. Also `.get()` might panic if the variable is unset, not sure if it might happen in our case. Probably we can accept that.